File: java/leetcode/src/Task297.java
import org.junit.jupiter.api.Test;

import java.util.*;

public class Task297 {
    //    public class Codec {
//
//        // Encodes a tree to a single string.
//        public String serialize(TreeNode root) {
//            if(root == null){
//                return "[]";
//            }
//            List<Integer> res = new ArrayList<>();
//
//            Queue<TreeNode> queue = new LinkedList<>();
//            queue.add(root);
//
//            while (!queue.isEmpty()){
//                TreeNode node = queue.poll();
//                if(node == null){
//                    res.add(null);
//                }else{
//                    res.add(node.val);
//                    queue.offer(node.left);
//                    queue.offer(node.right);
//                }
//
//            }
//            while(res.get(res.size()-1) == null){
//                res.remove(res.size()-1);
//            }
//            return res.toString();
//        }
//
//        // Decodes your encoded data to tree.
//        public TreeNode deserialize(String data) {
//            if(data.equals("[]")){
//                return null;
//            }
//            Integer[] array = Arrays.stream(data.substring(1, data.length() - 1).split(","))
//                    .map(String::trim).map(val->val.equals("null") ? null:Integer.valueOf(val)).toArray(Integer[]::new);
//
//            TreeNode root = new TreeNode(array[0]);
//            int cur = 1;
//
//            Queue<TreeNode> queue = new LinkedList<>();
//            queue.offer(root);
//
//            while(!queue.isEmpty()){
//                TreeNode node = queue.poll();
//                //left
//                if(cur < array.length){
//                    Integer left = array[cur++];
//                    if(left != null){
//                        TreeNode leftNode = new TreeNode(left);
//                        node.left = leftNode;
//                        queue.offer(leftNode);
//                    }
//                }
//
//                //right
//                if(cur < array.length){
//                    Integer right = array[cur++];
//                    if(right != null){
//                        TreeNode rightNode = new TreeNode(right);
//                        node.right = rightNode;
//                        queue.offer(rightNode);
//                    }
//                }
//            }
//
//            return root;
//        }
//
//
//
//    }
    public class Codec {
        // Decodes your encoded data to tree.
        public TreeNode deserialize(String data) {

            Queue<String> queue = new LinkedList<>(Arrays.asList(data.split(",")));

            TreeNode root = recursion(queue);

            return root;
        }

        private TreeNode recursion(Queue<String> queue) {
            String val = queue.poll();
            if(val.equals("nil")){
                return null;
            }else{
                TreeNode node = new TreeNode(Integer.valueOf(val));
                node.left = recursion(queue);
                node.right = recursion(queue);
                return  node;
            }

        }

        // Encodes a tree to a single string.
        public String serialize(TreeNode root) {

            StringBuilder sb = new StringBuilder();

            dfs(root, sb);

            return sb.toString();

        }

        private void dfs(TreeNode root, StringBuilder sb) {
            if(root == null){
                sb.append("nil,");
                return;
            }
            sb.append(root.val).append(",");

            dfs(root.left,sb);
            dfs(root.right,sb);
        }


    }

    @Test
    void test() {
        String data = "1,2,3,nil,nil,4,nil,nil,5,nil,nil,";
        Codec codec = new Codec();
        TreeNode root = codec.deserialize(data);
        String des = codec.serialize(root);

    }

}

File: py/Task297.py
# Definition for a binary tree node.
from queue import SimpleQueue


class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Codec:

    # # bfs
    # def serialize(self, root):
    #     l = []
    #     q = SimpleQueue()
    #     q.put(root)
    #     while not q.empty():
    #         node = q.get()
    #         if not node:
    #             l.append('N')
    #             continue
    #         l.append(str(node.val))
    #         q.put(node.left)
    #         q.put(node.right)
    #     return ','.join(l)

    # # bfs
    # def deserialize(self, data):
    #     data = data.split(',')
    #     if not data or data[0] == 'N':
    #         return None
    #     root = TreeNode(int(data[0]))
    #     q = SimpleQueue()
    #     q.put(root)
    #     pos = 1
    #     while not q.empty():
    #         node = q.get()
    #         left = data[pos]
    #         pos += 1
    #         right = data[pos]
    #         pos += 1
    #         if left != 'N':
    #             node.left = TreeNode(int(left))
    #             q.put(node.left)
    #         if right != 'N':
    #             node.right = TreeNode(int(right))
    #             q.put(node.right)
    #     return root

    # dfs
    def serialize(self, root):
        def dfs(node: TreeNode) -> str:
            if not node:
                return 'N'
            return str(node.val) + ','+dfs(node.left)+','+dfs(node.right)
        return dfs(root)
        # l = []

        # def dfs(node: TreeNode):
        #     if not node:
        #         l.append('N')
        #         return
        #     l.append(str(node.val))
        #     dfs(node.left)
        #     dfs(node.right)
        # dfs(root)
        # return ','.join(l)

    # dfs

    def deserialize(self, data):
        data = data.split(',')
        it = iter(data)

        def dfs() -> TreeNode:
            val = next(it)
            if val == 'N':
                return None
            node = TreeNode(int(val))
            node.left = dfs()
            node.right = dfs()
            return node
        return dfs()


codec = Codec()
s = "1,2,N,N,3,4,N,N,5,N,N"
root = codec.deserialize(s)
s1 = codec.serialize(root)
print(s1)
